container/container.go

```diff
@@ -350,19 +350,20 @@ func (container *Container) GetExecIDs() []string {
 
 // Attach connects to the container's TTY, delegating to standard
 // streams or websockets depending on the configuration.
-func (container *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, keys []byte) chan error {
+func (container *Container) Attach(stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, keys []byte) (chan bool, chan error) {
 	ctx := container.InitAttachContext()
 	return AttachStreams(ctx, container.StreamConfig, container.Config.OpenStdin, container.Config.StdinOnce, container.Config.Tty, stdin, stdout, stderr, keys)
 }
 
 // AttachStreams connects streams to a TTY.
 // Used by exec too. Should this move somewhere else?
-func AttachStreams(ctx context.Context, streamConfig *runconfig.StreamConfig, openStdin, stdinOnce, tty bool, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, keys []byte) chan error {
+func AttachStreams(ctx context.Context, streamConfig *runconfig.StreamConfig, openStdin, stdinOnce, tty bool, stdin io.ReadCloser, stdout io.Writer, stderr io.Writer, keys []byte) (chan bool, chan error) {
 	var (
 		cStdout, cStderr io.ReadCloser
 		cStdin           io.WriteCloser
 		wg               sync.WaitGroup
 		errors           = make(chan error, 3)
+		detachChan       = make(chan bool, 1)
```

Comment:
Ok, I don't like channel here :/ And detach info is not really asynchronous to an execution of promise.Go. I'd prefer to use probably a special error from AttachStreams, then it will be less disruptive for API. We can see that stdin copy and part of promise after <-done is synchronous. So you can just create bool variable, set it in stdin goroutine and read it in promise and if it's true - return special error, like `errDetached`.